pkg/vtpm/swtpm-vtpm/src/main.go

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) 2024 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net"
+	"os"
+	"os/exec"
+	"os/signal"
+	"strings"
+	"syscall"
+
+	"github.com/google/go-tpm/tpm2"
+	etpm "github.com/lf-edge/eve/pkg/pillar/evetpm"
+)
+
+const (
+	swtpmPath = "/usr/bin/swtpm"
+	// FIX-ME : use constanst from eve types, in another PR
+	// TpmdControlSocket is UDS to aks vtpmd to luanch SWTP instances for VMS
+	TpmdControlSocket = "/run/swtpm/tpmlaunchd"
+	// SwtpmSocketPath is the prefix for the SWTPM socket
+	SwtpmSocketPath = "/run/swtpm/%s.sock"
+	// SwtpmPidPath is the prefix for the SWTPM pid file
+	SwtpmPidPath = "/run/swtpm/%s.pid"
+
+	stateEncryptionKey = "/run/swtpm/binkey"
+	swtpmLogPath       = "/run/swtpm/%s.log"
+	swtpmStatePath     = "/persist/swtpm/tpm-state-%s"
+	maxInstances       = 10
+)
+
+var liveInstances = 0
+
+func makeDirs(dir string) error {
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		if err := os.MkdirAll(dir, 0755); err != nil {
+			return fmt.Errorf("Failed to create directory: %v", err)
+		}
+	} else if err != nil {
+		return fmt.Errorf("Failed to check directory: %v", err)
+	} else {
+		// dir exists, make sure it has the right permissions
+		if err := os.Chmod(dir, 0755); err != nil {
+			return fmt.Errorf("Failed to set permissions for  directory: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func runVirtualTpmInstance(id string) error {
+	statePath := fmt.Sprintf(swtpmStatePath, id)
+	logPath := fmt.Sprintf(swtpmLogPath, id)
+	sockPath := fmt.Sprintf(SwtpmSocketPath, id)
+	pidPath := fmt.Sprintf(SwtpmPidPath, id)
+
+	if err := makeDirs(statePath); err != nil {
+		return fmt.Errorf("Failed to create vtpm state directory: %v", err)
+	}
+
+	_, err := os.Stat(etpm.TpmDevicePath)
+	if err != nil {
+		log.Println("TPM is not available, running swtpm without state encryption!")
+
+		cmd := exec.Command(swtpmPath, "socket", "--tpm2",
+			"--tpmstate", "dir="+statePath,
+			"--ctrl", "type=unixio,path="+sockPath+",terminate",
+			"--log", "file="+logPath+",level=20", // FIX-ME: lower the log level, or get rid of it
+			"--pid", "file="+pidPath,
+			"--daemon")
+		if err := cmd.Run(); err != nil {
+			return fmt.Errorf("Failed to run swtpm: %v", err)
+		}
+	} else {
+		log.Println("TPM is available, running swtpm with state encryption!")
+		rw, err := tpm2.OpenTPM(etpm.TpmDevicePath)
+		if err != nil {
+			return fmt.Errorf("OpenTPM failed with err: %v", err)
+		}
+		defer rw.Close()
+
+		key, err := etpm.UnsealDiskKey(etpm.DiskKeySealingPCRs)
+		if err != nil {
+			return fmt.Errorf("Unseal operation failed with err: %v", err)
+		}
+
+		if err := ioutil.WriteFile(stateEncryptionKey, key, 0644); err != nil {
```

Comment:
this is deprecated: https://pkg.go.dev/io/ioutil#WriteFile
```suggestion
		if err := os.WriteFile(stateEncryptionKey, key, 0644); err != nil {
```